src/main/scala/rise/core/types/Type.scala

```diff
@@ -2,8 +2,88 @@ package rise.core.types
 
 import arithexpr.arithmetic.RangeAdd
 import rise.core._
+import util.PatternMatching
 
-sealed trait Type
+object alphaEquiv {
+  type Env = List[(Kind.Identifier , Kind.Identifier)]
+
+  def equivNat(env : Env) : Nat => Nat => Boolean = a => b => {
+    val natEnv = env.collect {case (i : NatIdentifier, n : Nat) => (i , n)}
+    // substitutes elements on the left with elements on the right
+    substitute.natsInNat(natEnv.toMap, a) == b
+  }
+
+  def makeExplicit[K <: Kind.Identifier] : K => K = {
+    case t : DataTypeIdentifier => t.asExplicit.asInstanceOf[K]
+    case t => t
+  }
+
+  def equiv : Type => Type => Boolean = equiv(Nil)
+
+  /** Alpha equivalence on types
+    * Datatype identifier explicitness is ignored.
+    * Short circuits in the event of syntactic equality.
```

Comment:
is this short-circuit not more costly in most cases (we might have to do two traversals instead of one)?

Reply:
I need to do some benchmarking, but I suspect Scala's equals is heavily optimized, like, maybe it's able to see that it hasn't been overwritten and use a direct memory comparison or something.